go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -149,6 +156,73 @@ func delPidfile(pidfile string) {
 	}
 }
 
+// ovnkubeControllerSyncPostBootFilePath is located in storage that does not persist following a reboot. ovn-controller start
+// is predicated on the existence of this file.
+const ovnkubeControllerSyncPostBootFilePath = "/tmp/ovnkube-controller-sync-post-boot"
+
+func ensureOVNKubeControllerSyncPostBootFile() error {
+	if err := os.WriteFile(ovnkubeControllerSyncPostBootFilePath, []byte(time.Now().String()), 0o644); err != nil {
+		return fmt.Errorf("failed to write to file at path %q: %v", ovnkubeControllerSyncPostBootFilePath, err)
+	}
+	return nil
+}
+
+// doesOVNKubeControllerSyncPostBootFileExist checks if a sentinel file exists. File will not exist post Node restart. OVNKube controller creates the file post sync and after a reboot.
+func doesOVNKubeControllerSyncPostBootFileExist() (bool, error) {
+	exists := false
+	_, err := os.Stat(ovnkubeControllerSyncPostBootFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return exists, fmt.Errorf("failed to stat file at path %q: %w", ovnkubeControllerSyncPostBootFilePath, err)
+		}
+		exists = true
+	}
+	return exists, nil
+}
+
+// waitUntilNorthdSyncOnce ensures northd has sync'd at least once by increments nb_cfg value in NB DB and waiting
+// for northd to copy it to SB DB. Poll SB DB until context is cancelled.
+func waitUntilNorthdSyncOnce(ctx context.Context, nbClient, sbClient client.Client) error {
+	// 1. Get value of nb_cfg
+	// 2. Increment value of nb_cfg
+	// 3. Wait until value appears in SB DB after northd copies it.
+	nbGlobal := &nbdb.NBGlobal{}
+	nbGlobal, err := libovsdbops.GetNBGlobal(nbClient, nbGlobal)
+	if err != nil {
+		return fmt.Errorf("failed to find OVN Northbound NB_Global table"+
+			" entry: %v", err)
+	}
+	// increment nb_cfg value by 1. When northd consumes updates from NB DB, it will copy this value to SB DBs SB_Global table nb_cfg field.
+	ops, err := nbClient.Where(nbGlobal).Mutate(nbGlobal, model.Mutation{
+		Field:   &nbGlobal.NbCfg,
+		Mutator: ovsdb.MutateOperationAdd,
+		Value:   1,
```

Comment:
nit; may add comment that no other updates happening for `nbGlobal.NbCfg` field at this time.

Reply:
its actually fine if theres updates at the same time. May I ask why it wouldnt be fine?

Reply:
that may be because check on `sbGlobal.NbCfg >= expectedNbCfgValue` caused by other update.

Reply:
nb_cfg is only ever increment by any ovnk code.

We dont care above others incrementing it because when we get the nb_cfg value initally, we know the NB DB is HOT with all the data we want to copy to SB DB, so if we see that value + 1 or greater in SB DB, then we know whatever was in the NB DB for the initial nb_cfg value is in SB DB. Make sense?